File: sort-mapper-example-java/src/main/java/kh/farrukh/sortmapperexamplejava/teacher/TeacherMapper.java
package kh.farrukh.sortmapperexamplejava.teacher;

import kh.farrukh.sortmapperexamplejava.teacher.model.TeacherCreateRequestDTO;
import kh.farrukh.sortmapperexamplejava.teacher.model.TeacherDetailsResponseDTO;
import kh.farrukh.sortmapperexamplejava.teacher.model.TeacherResponseDTO;
import kh.farrukh.sortmapperexamplejava.teacher.model.TeacherUpdateRequestDTO;
import org.mapstruct.Mapper;
import org.mapstruct.MappingConstants.ComponentModel;
import org.mapstruct.MappingTarget;

@Mapper(componentModel = ComponentModel.SPRING)
public interface TeacherMapper {

  TeacherResponseDTO toResponseDTO(TeacherEntity entity);

  TeacherDetailsResponseDTO toDetailsResponseDTO(TeacherEntity entity);

  TeacherEntity toEntity(TeacherCreateRequestDTO requestDTO);

  void update(@MappingTarget TeacherEntity entity, TeacherUpdateRequestDTO requestDTO);

}

File: sort-mapper-example/src/main/kotlin/kh/farrukh/sortmapperexample/teacher/TeacherMapper.kt
package kh.farrukh.sortmapperexample.teacher

import kh.farrukh.sortmapperexample.teacher.model.TeacherCreateRequestDTO
import kh.farrukh.sortmapperexample.teacher.model.TeacherDetailsResponseDTO
import kh.farrukh.sortmapperexample.teacher.model.TeacherResponseDTO
import kh.farrukh.sortmapperexample.teacher.model.TeacherUpdateRequestDTO

fun TeacherEntity.toResponseDTO() = TeacherResponseDTO(
    id = id!!,
    firstName = firstName,
    lastName = lastName,
    middleName = middleName,
    age = age
)

fun TeacherCreateRequestDTO.toEntity() = TeacherEntity(
    firstName = firstName,
    lastName = lastName,
    middleName = middleName,
    age = age
)

fun TeacherEntity.toDetailsResponseDTO() = TeacherDetailsResponseDTO(
    id = id!!,
    firstName = firstName,
    lastName = lastName,
    middleName = middleName,
    age = age
)

fun TeacherEntity.update(requestDTO: TeacherUpdateRequestDTO) {
    firstName = requestDTO.firstName
    lastName = requestDTO.lastName
    middleName = requestDTO.middleName
    age = requestDTO.age
}